recipes/obake/meta.yaml

```diff
@@ -0,0 +1,53 @@
+{% set version = "0.1" %}
+
+package:
+  name: obake
+  version: {{ version }}
+
+source:
+  url: https://github.com/bluescarni/obake/archive/v{{ version }}.tar.gz
+  sha256: 946de817940421d8af8c614c289b74dc6e221ea803fc2b3dacc30e63d7dc3246
+
+build:
+  number: 0
+  run_exports:
+    # SO name changes at minor rev bumps.
+    - {{ pin_subpackage('obake', max_pin='x.x') }}
+  skip: true   # [win and vc<14]
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - cmake
+    - clang      # [win]
+    - ninja      # [win]
+  host:
+    - abseil-cpp
+    - boost-cpp
+    - mppp
+    - tbb
+    - tbb-devel
+    - backtrace    # [linux]
+  run:
+    - abseil-cpp
+    - boost-cpp
+    - {{ pin_compatible("mppp") }}
+    - tbb
+    - tbb-devel
```

Comment:
Is this needed at runtime?

Reply:
I admit I am confused about the whole run vs host dependency difference.

The library has a public dependence on boost, mppp and tbb, meaning that its public headers include various headers from these libraries, and meaning that if you want to use the library you will have to link to libmp++, libtbb, etc. Does this count as runtime usage? Or is this part of the host dependencies?

Reply:
`host` should include packages with static libraries, shared, headers, etc. Run should include packages that are needed at runtime.

As an example, assume you use this package in a python extension and that package pulls in obake because of the shared library, but it doesn't need the headers, static libraries etc.

You have 2 options.

1. You can have `obake` which pulls in boost headers, etc and have `libobake` be just the shared libraries. Then you need to add `run_exports` on `libobake` and `tbb` in `obake`, because downstream packages need to link both `libobake` and `tbb`

2. Remove boost-cpp and tbb-devel from `obake` and add them in downstreams.

Reply:
I think I need all these libraries as run dependencies, as they all need to be present if a package wants to use the functionality in obake. If they are not, and one tries, e.g., to include a obake header, the compiler will error out due to the missing tbb, boost, abseil, etc. Are you expecting that a package depending on obake should depend manually on obake's dependencies or am I misunderstanding something?

obake is really a header-only library with a very small compiled part. How would a header-only C++ library fit in the ``run``/``host`` scheme?

Reply:
Then option 1 is the way to go

Reply:
symengine does the same for gmp/mpfr:

https://github.com/conda-forge/symengine-feedstock/blob/master/recipe/meta.yaml

How is this situation any different?

Reply:
Or xtensor/xtl:

https://github.com/conda-forge/xtensor-feedstock/blob/master/recipe/meta.yaml#L23

Reply:
symengine doesn't have `llvmdev` in run, because it's linked in statically. (Currently, llvm headers are not used in symengine headers, but earlier symengine headers used them and still we didn't have `llvmdev` in `run` because `llvmdev` also pulls in the static libraries)

Reply:
> Then option 1 is the way to go

Does this mean I need to have 2 packages?

Reply:
You can use `outputs` in conda-build.

Reply:
> symengine doesn't have `llvmdev` in run, because it's linked in statically. (Currently, llvm headers are not used in symengine headers, but earlier symengine headers used them and still we didn't have `llvmdev` in `run` because `llvmdev` also pulls in the static libraries)

Sure ok that makes sense. But in symengine you are both linking to ``gmp.so`` and including ``gmp.h`` in the symengine public headers, and thus ``gmp`` goes both in ``run`` and ``host``. Is that correct?

Reply:
Yes. If symengine used only `gmp.h` (like obake does boost) and gmp was a heavy dependency, I would have split symengine packages.

Reply:
> we didn't have `llvmdev` in `run` because `llvmdev` also pulls in the static libraries

What is the problem if ``llvmdev`` as a dep pulls in static libraries?

Reply:
When I do `conda install python-symengine`, I don't want it to pull in `llvmdev` which is hundreds of megabytes in size. `llvmdev` is already statically linked in the `symengine` shared library.

Reply:
Any pointer on how to use ``outputs``? It looks significantly more complicated than anything I have written for conda.

Or maybe better to just vendor obake for now? We gonna start depend on it in other conda packages soon, that might be the easiest way for us to proceed.

Reply:
https://github.com/conda-forge/openblas-feedstock/blob/master/recipe/meta.yaml

Reply:
Ok I'll drop it then, we'll include a copy of obake in another recipe.

Reply:
It's not that hard. I can help if you run into any trouble.

Reply:
From the openblas yaml above I cannot really figure out what's going on. Am I supposed to list by hand in the recipe the files that need to go in one package vs the files that need to go in another? But then the tbb yaml seems to be doing things in a different fashion (calls some ``{{ build }}`` script with different arguments)?

Reply:
And what is going on with the nomenclature? In the openblas yaml there's ``openblas`` vs ``libopenblas``, while in the tbb recipe there's ``tbb`` vs ``tbb-devel``. Is this the same idea or something different? Is the idea behind this split similar to the ``-devel`` split in some linux distro or is it something else?

Reply:
> Am I supposed to list by hand in the recipe the files that need to go in one package vs the files that need to go in another?

In that case, you do `make install` in the top level build script and then list the files in each output.

As in the case of `tbb`, you can do `make` in the top level build script and then in the outputs, only install the files relating to the output. (You can also do make install and remove files not belonging to that output.

> And what is going on with the nomenclature? 

It's adhoc and not standardized. openblas has historical reasons, because `openblas` used to have everything (shared libraries, static libraries, headers) and can't be changed. Yes `tbb-devel` vs `tbb` is like the split in linux distros.

Reply:
I find the openblas recipe confusing and not particularly useful as a general example.

For instance: am I supposed to duplicate all the ``skip`` statements in the various outputs or is it enough to have it top level only? Same with the build requirements, is it really necessary to repeat everywhere we need a C/fortran compiler for the build?

It also does not really clarify how to partition the dependencies into run/host for the various subpackages, because openblas does not really depend on anything.